pkg/scheduler/preemption/preemption.go

```diff
@@ -222,14 +239,101 @@ func minimalPreemptions(wl *workload.Info, assignment flavorassigner.Assignment,
 			snapshot.RemoveWorkload(targets[i])
 		}
 	}
-	// Reset changes to the snapshot.
+	return targets
+}
+
+func restoreSnapshot(snapshot *cache.Snapshot, targets []*workload.Info) {
 	for _, t := range targets {
 		snapshot.AddWorkload(t)
 	}
+}
 
+func fairPreemptions(wl *workload.Info, assignment flavorassigner.Assignment, snapshot *cache.Snapshot, resPerFlv resourcesPerFlavor, candidates []*workload.Info, allowBorrowingBelowPriority *int32) []*workload.Info {
+	cqHeap := heap.New(
+		func(c *candidateCQ) string {
+			return c.cq.Name
+		},
+		func(c1, c2 *candidateCQ) bool {
+			return c1.share > c2.share
+		},
+	)
+	for _, cand := range candidates {
+		candCQ := cqHeap.GetByKey(cand.ClusterQueue)
+		if candCQ == nil {
+			cq := snapshot.ClusterQueues[cand.ClusterQueue]
+			share, _ := cq.DominantResourceShare(nil)
+			candCQ = &candidateCQ{
+				cq:        cq,
+				share:     share,
+				workloads: []*workload.Info{cand},
+			}
+			_ = cqHeap.PushIfNotPresent(candCQ)
+		} else {
+			candCQ.workloads = append(candCQ.workloads, cand)
+		}
+	}
+
+	nominatedCQ := snapshot.ClusterQueues[wl.ClusterQueue]
+	newNominatedShareValue, _ := nominatedCQ.DominantResourceShare(wl)
+	wlReq := totalRequestsForAssignment(wl, assignment)
+	var targets []*workload.Info
+	fits := false
+	for cqHeap.Len() > 0 && !fits {
+		candCQ := cqHeap.Pop()
+
+		if candCQ.cq == nominatedCQ {
+			candWl := candCQ.workloads[0]
+			snapshot.RemoveWorkload(candWl)
+			targets = append(targets, candWl)
+			if workloadFits(wlReq, nominatedCQ, true) {
+				fits = true
+				break
+			}
+			newNominatedShareValue, _ = nominatedCQ.DominantResourceShare(wl)
+			candCQ.workloads = candCQ.workloads[1:]
+			if len(candCQ.workloads) > 0 {
+				candCQ.share, _ = candCQ.cq.DominantResourceShare(nil)
+				cqHeap.PushIfNotPresent(candCQ)
+			}
+			continue
+		}
+
+		for i, candWl := range candCQ.workloads {
```

Comment:
Similar question as above, we seem to rely on the order of workloads below, is it ordered?